docs/source/concepts/time.rst

```diff
@@ -0,0 +1,90 @@
+and defines the value of all :ref:`getTime <daml-ref-gettime>` calls in the given transaction.
+The ledger time is assigned by the participant as part of the DAML command interpretation.
+
+The ledger time of valid transaction ``TX`` must fullfil the following rules:
+
+#. **Causal monotonicity**: for any action (create, exercise, fetch, lookup) in ``TX``
+   on a contract ``C``, ``lt_TX >= lt_C``,
+   where ``lt_C`` is the ledger time of the transaction that created ``C``.
+
+#. **Skew**: ``rt_TX - skew_min <= lt_TX <= rt_TX + skew_max``,
+   where ``skew_min`` and ``skew_max`` are parameters defined by the ledger.
+
+Apart from that, no other guarantees are given on the ledger time.
+Time has therefore to be considered slightly fuzzy in DAML, with the fuzziness depending on the skew parameters.
+DAML applications should not interpret the value returned by :ref:`getTime <daml-ref-gettime>` as a precise timestamp.
+
+
+Record time
+***********
+
+The *record time* ``rt_TX`` is another property of a transaction.
+It is timestamp with microsecond resolution,
+and is assigned by the ledger when the transaction is recorded on the ledger.
+
+The DAML ledger model does not prescribe how the record time is assigned.
+
+Ledger time model
+*****************
+
+The *ledger time model* is the set of parameters used in the assignment and validation of ledger time.
+It consists of the following:
+
+#. ``skew_min`` and ``skew_max``, the bounds on the difference between ``lt_TX`` and ``rt_TX``.
+
+#. ``transaction_latency``, the average duration from the time a transaction is submitted from a participant to the ledger
+   until the transaction is recorded.
+
+
+The ledger time model is part of the ledger configuration and can be changed by ledger operators through the
+:ref:`SetTimeModel <com.daml.ledger.api.v1.admin.SetTimeModelRequest>` config management API.
+
+Assigning ledger time
+*********************
+
+The ledger time is assigned automatically by the participant.
+In most cases, DAML applications will not need to worry about ledger time and record time at all.
+
+For reference, this section describes the details of how the ledger time is assigned. 
+
+#. When submitting commands over the ledger API,
+   users can optionally specify a ``min_ledger_time_rel`` or ``min_ledger_time_abs`` argument.
+   This defines a lower bound for the ledger time in relative and absolute terms, respectively.
+
+#. The ledger time is set to the highest of the following values:
+
+   #. ``max(lt_C_1, ..., lt_C_n)``, the maximum ledger time of all contracts used by the given transaction
+   #. ``t_p``, the local time on the participant
+   #. ``t_p + min_ledger_time_rel``, if ``min_ledger_time_rel`` is given
+   #. ``min_ledger_time_abs``, if ``min_ledger_time_abs`` is given
+
+#. If the set of contracts used by the transaction depends on time,
+   then the command interpretation may need to be repeated until a suitable ledger time is found.
```

Comment:
What is used for the first interpretation which you need to determine the contracts used by the transaction? Do we first calculate the time with the set of contracts used by the transaction being empty?

Reply:
Yes, that's basically what happens.

This section does not describe the actual algorithm, only the properties of the algorithm. The actual algorithm is:

1. Pick a ledger time using the current local time and the `min_ledger_time_*` parameters
2. Interpret the command. During interpretation, track whether the ledger time was used.
3. Get the maximum ledger time of all contracts used by the resulting transaction. If this fails (because one of the used contracts was archived between step 2 and 3), abort and fail the command submission.
  3.1. If the transaction does not violate causal monotonicity, finish.
  3.2. If the transaction violates causal monotonicity and the transaction depends on time, advance the ledger time and go back to step 2. After 3 iterations, abort and fail the command submission.
  3.3. If the transaction violates causal monotonicity and the transaction does *not* depend on time, modify the ledger time of the transaction and finish.

Do you think the reader would care about these details?

Reply:
I think it’s fine to omit those details but it might be helpful to clarify that the description is not intended to reflect the full algorithm but only a high-level overview. Maybe just rephrasing the point here as follows helps

> Since the set of commands used by given transaction can depend on the chosen time, the above process might need to be repeated until a suitable ledger time is found.

Reply:
Also: is there a bound on the number of the repetitions?

Reply:
Yes, the maximum number of iterations is 3. I have added it  to the text.